### PR TITLE
ARROW-5558: [C++] Support Array::View on arrays with non-zero offset

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

File: cpp/src/arrow/array_view_test.cc [hunk 1]
```diff
@@ -406,18 +406,32 @@ TEST(TestArrayView, ExtensionType) {
...
+  ASSERT_OK(ListArray::FromArrays(*list_offsets, *view_values->Slice(2),
+                                  default_memory_pool(), &expected));
+  ASSERT_OK(arr->Validate());
+  CheckView(arr->Slice(1), expected->Slice(1));
+
+  // Be extra paranoid about checking offsets
+  std::shared_ptr<Array> result;
+  ASSERT_OK(arr->Slice(1)->View(expected->type(), &result));
+  ASSERT_EQ(1, result->offset());
+  ASSERT_EQ(2, static_cast<const ListArray&>(*result).values()->offset());
```
Comment (bkietz) [inline cpp/src/arrow/array_view_test.cc]:
```suggestion
  ASSERT_EQ(2, static_cast<const ListArray&>(*result).values()->offset());
  CheckView(arr->Slice(1), result);
```

**Review by bkietz — APPROVED** [review]
LGTM

**Comment by bkietz** [comment]
Travis failure was a timeout: https://travis-ci.org/apache/arrow/jobs/581008865#L2431

File: cpp/src/arrow/array_view_test.cc (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/array_view_test.cc]: This isn't needed

**Pull request closed without merging**